File: 001_010/AddTwoNumbers.py
# 002. Add Two Numbers

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # 计算头结点的和，将结果赋值给了，并计算carry
        sum = l1.val + l2.val
        ans = sum % 10
        carry = sum // 10
        res = ListNode(ans)
        # 遍历了l1和l2，其中用next是因为在结束时，我们要将l1, l2指向最后一个非null结点，这样可以在后面继续添加1
        l3 = res
        while l1.next and l2.next:
            l1 = l1.next
            l2 = l2.next
            sum = l1.val + l2.val + carry
            ans = sum % 10
            carry = sum // 10
            l3.next = ListNode(ans)
            l3 = l3.next
        # l1的长度大于l2，直接在l1上修改值
        while l1.next:
            l1 = l1.next
            sum = l1.val + carry
            ans = sum % 10
            carry = sum // 10
            l3.next = ListNode(ans)
            l3 = l3.next
        # l2的长度大于l1，直接在l1上添加结点
        while l2.next:
            l2 = l2.next
            sum = l2.val + carry
            ans = sum % 10
            carry = sum // 10
            l3.next = ListNode(ans)
            l3 = l3.next

        if not l1.next and not l2.next and carry == 1:
            l3.next = ListNode(1)

        return res

File: go_solution/001_010/addTwoNumbers/addTwoNumbers.go
package addTwoNumbers

// 002. Add Two Numbers

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	tmp := 0
	head := &ListNode{Val: 0, Next: nil} // dummy node to track the head of the result
	node := head
	for l1 != nil || l2 != nil || tmp != 0 {
		if l1 != nil {
			tmp = tmp + l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			tmp = tmp + l2.Val
			l2 = l2.Next
		}
		node.Next = &ListNode{
			Val:  tmp % 10,
			Next: nil,
		}
		node = node.Next
		tmp = tmp / 10
	}
	return head.Next
}
